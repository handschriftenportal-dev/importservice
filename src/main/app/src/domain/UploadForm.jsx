import React from "react";
import {AuthenticationContext} from '../infrastructure/AuthenticationContext'
import {
  Button,
  ButtonGroup,
  Grid,
  Hidden,
  Input,
  InputLabel,
  Typography
} from '@material-ui/core';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';
import Box from "@material-ui/core/Box";
import IconButton from "@material-ui/core/IconButton";
import ArrowUpwardIcon from '@material-ui/icons/ArrowUpward';
import TextField from "@material-ui/core/TextField";
import {FormattedMessage} from "react-intl";

class UploadForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      title: "XML Datei Upload",
      datei: React.createRef(),
      invalid: true,
      dateiKey: new Date(),
      apiEndpoint: this.props.endpoint,
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleOnChange = this.handleOnChange.bind(this)
    this.handleOnChangeForReset = this.handleOnChangeForReset.bind(this)
  }

  handleOnChange(event) {
    if (event.target.value) {
      this.setState({invalid: false})
    }
  }

  handleOnChangeForReset(event) {

    console.log("Reset Form")
    this.setState({datei: React.createRef(), invalid: true})
  }

  async handleSubmit(event) {

    event.preventDefault();

    if (!this.state.datei.current || !this.state.datei.current.files[0]) {
      alert("Bitte Datei auswählen.")

      this.state.datei.current.focus();
    } else {
      console.log(
          'Starting upload of file ' + this.state.datei.current.files[0].name
          + " to API " + this.state.apiEndpoint);
      console.log(this.state.datei.current.files[0])

      await this.upload(this.state.datei.current.files[0])

      this.setState(
          {datei: React.createRef(), invalid: true, dateiKey: new Date()})

      this.props.tableUpdate();
    }
  }

  async upload(file) {

    const formData = new FormData();

    formData.append('datei', file);

    await window.fetch(this.state.apiEndpoint, {
      method: 'POST',
      headers: {'Accept-Encoding': 'gzip, deflate'},
      body: formData,
    }).then((response) => {
      console.log("Response of uploading file")
      console.log(response)
    }).catch(function (error) {
      console.log("Error occurred during upload")
      console.log(error)
    });
  }

  componentDidMount() {
    console.log("Mount UploadForm")
  }

  componentWillUnmount() {
    console.log("Umount UploadForm")
  }

  render() {
    return <Grid container justify="center" alignItems="center"
                 direction="column">
      <Hidden>
        <Grid item md={1}></Grid>
      </Hidden>
      <Grid item xs={12} md={10}>
        <Typography
            id="uploadheadline" variant="h5"
            component="h6">{<FormattedMessage id="form.welcome"/>}
          {' ' + this.context.username}</Typography>
        <form id="uploadform" onSubmit={this.handleSubmit} title="uploadform"
              aria-invalid={this.state.invalid} aria-label={'uploadform'}>
          <Box m={2}>
            <Box component="div" visibility="hidden">
              <InputLabel htmlFor="uploadFile"
                          aria-label={'uploadfilelabel'}>{this.state.title}</InputLabel>
            </Box>
            <IconButton
                variant="contained" aria-label="button"
                component="label" style={{width: "300px"}}
            ><ArrowUpwardIcon
                aria-label="button">UpArrowButton</ArrowUpwardIcon>
              <Input key={this.state.dateiKey} type="file"
                     name="datei"
                     onChange={this.handleOnChange}
                     placeholder=""
                     style={{display: "none"}}
                     id="uploadFile" inputRef={this.state.datei}
                     aria-label="uploadinputformular"/>
              <TextField id="standard-basic" variant="outlined" fullWidth
                         label={!this.state.datei.current
                         || !this.state.datei.current.files[0] ? ''
                             : this.state.datei.current.files[0].name}/>
            </IconButton>
          </Box>
          <Box m={2}>
            <div>
              <ButtonGroup aria-label="outlined secondary button group">
                <Button id="submitupload" aria-label="hinzufügen"
                        type="submit"
                        variant="outlined"
                        color="primary"
                        startIcon={<CloudUploadIcon/>}>
                  <FormattedMessage id="form.hinzufuegen"/>
                </Button>
                <Button variant="outlined" color="secondary" type="reset"
                        aria-label="zurücksetzen"
                        onClick={this.handleOnChangeForReset}>
                  <FormattedMessage id="form.zuruecksetzen"/>
                </Button>
              </ButtonGroup>
            </div>
          </Box>
        </form>
      </Grid>
      <Hidden>
        <Grid item md={1}></Grid>
      </Hidden>
    </Grid>
  }
}

UploadForm.contextType = AuthenticationContext;

export default UploadForm
