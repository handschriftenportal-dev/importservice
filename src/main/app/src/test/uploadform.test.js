import React from "react";
import {unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import UploadForm from "../domain/UploadForm";
import createComponentWithIntl from './createComponentWithIntl';

describe('Import Upload Form Testsuite', () => {

  let container = null;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it("Test UploadFrom Expected Content", () => {

    let form = createComponentWithIntl(<UploadForm id="uploadform"
                                                   tableUpdate={() => {
                                                   }}
                                                   endpoint="/rest/import"></UploadForm>),
        container;

    //expect(form.exists("#uploadheadline")).to.equal(true);

    expect(form.root.findAllByType("button")).toHaveLength(2);

  })

  it('UploadForm Snapshot Test', () => {
    const tree =
        createComponentWithIntl(<UploadForm id="uploadform"
                                            tableUpdate={() => {
                                            }}
                                            endpoint="http://b-dev1047.pk.de:9296/rest/import"></UploadForm>)
        .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('UploadForm Click Submit Button', () => {
    const onClick = jest.fn();
    const jsdomAlert = window.alert;
    window.alert = () => {
    };

    act(() => {
      createComponentWithIntl(<UploadForm id="uploadform"
                                          tableUpdate={onClick}
                                          endpoint="http://b-dev1047.pk.de:9296/rest/import"></UploadForm>),
          container;
    });

    /* const button = document.querySelector("#submitupload");

     act(() => {
       button.dispatchEvent(new MouseEvent("click", {bubbles: true}));
     });

     expect(onClick).toHaveBeenCalledTimes(0);*/

  });

});
