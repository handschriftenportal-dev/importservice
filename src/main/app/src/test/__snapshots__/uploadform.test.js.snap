// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import Upload Form Testsuite UploadForm Snapshot Test 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-10"
  >
    <h6
      className="MuiTypography-root MuiTypography-h5"
      id="uploadheadline"
    >
      Willkommen lieber
       Gast
    </h6>
    <form
      aria-invalid={true}
      aria-label="uploadform"
      id="uploadform"
      onSubmit={[Function]}
      title="uploadform"
    >
      <div
        className="MuiBox-root MuiBox-root-297"
      >
        <div
          className="MuiBox-root MuiBox-root-298"
        >
          <label
            aria-label="uploadfilelabel"
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
            htmlFor="uploadFile"
          >
            XML Datei Upload
          </label>
        </div>
        <label
          aria-disabled={false}
          aria-label="button"
          className="MuiButtonBase-root MuiIconButton-root"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          style={
            Object {
              "width": "300px",
            }
          }
          tabIndex={0}
          variant="contained"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              aria-label="button"
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
              />
            </svg>
            <div
              aria-label="uploadinputformular"
              className="MuiInputBase-root MuiInput-root MuiInput-underline"
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
            >
              <input
                className="MuiInputBase-input MuiInput-input"
                id="uploadFile"
                name="datei"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                type="file"
              />
            </div>
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  id="standard-basic"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-230 MuiOutlinedInput-notchedOutline"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="PrivateNotchedOutline-legend-231"
                    style={
                      Object {
                        "width": 0.01,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
              </div>
            </div>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </label>
      </div>
      <div
        className="MuiBox-root MuiBox-root-299"
      >
        <div>
          <div
            aria-label="outlined secondary button group"
            className="MuiButtonGroup-root"
            role="group"
          >
            <button
              aria-label="hinzuf端gen"
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-outlinedPrimary"
              disabled={false}
              id="submitupload"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="submit"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
                    />
                  </svg>
                </span>
                Hinzuf端gen
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="zur端cksetzen"
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-outlinedSecondary"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="reset"
            >
              <span
                className="MuiButton-label"
              >
                Zur端cksetzen
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
